docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1766 @@
+   `with_option` internally but provide two additional features:
+   1. They validate that the option is indeed a stopping/convergence criterion.
+   1. They allow to omit the `convergence_` or `stopping_` at the beginning of the
+      option name and can thus reduce repetition in the option names.
+1. As before, the user can pass a global set of options to `maximize` or `minimize`. We
+   continue to support option dictionaries but also allow `AlgorithmOption` objects that
+   enable better autocomplete and immutability. We can construct them using a similar
+   pre-commit hook approach as discussed in [algorithm selection](algorithm-selection).
+   Global options override the options that were directly passed to an optimizer. For
+   consistency, `AlgorithmOptions` can offer the `with_stopping`, `with_convergence` and
+   `with_option` copy-constructors, so users can modify options safely. Probably, this
+   approach should be featured less prominently in the documentation as it offers no
+   guarantees that the specified options are compatible with the selected algorithm.
+
+The previous example continues to work. Examples of the new possibilities are:
+
+```python
+# configured algorithm
+algo = em.algorithms.scipy_lbfgsb(
+    stopping_max_iterations=1000,
+    stopping_max_criterion_evaluations=1500,
+    convergence_relative_criterion_tolerance=1e-6,
+)
+minimize(
+    # ...
+    algorithm=algo,
+    # ...
+)
+```
+
+```python
+# using copy constructors for better namespaces
+algo = (
+    em.algorithms.scipy_lbfgsb()
+    .with_stopping(
+        max_iterations=1000,
+        max_criterion_evaluations=1500,
+    )
+    .with_convergence(
+        relative_criterion_tolerance=1e-6,
+    )
+)
+
+minimize(
+    # ...
+    algorithm=algo,
+    # ...
+)
+```
+
+```python
+# using copy constructors to create variants
+base_algo = em.algorithms.fides(stopping_max_iterations=1000)
+algorithms = [base_algo.with_option(initial_radius=r) for r in [0.1, 0.2, 0.5]]
+
+for algo in algorithms:
+    minimize(
+        # ...
+        algorithm=algo,
+        # ...
+    )
+```
+
+```python
+# option object
+options = em.AlgorithmOptions(
+    stopping_max_iterations=1000,
+    stopping_max_criterion_evaluations=1500,
+    convergence_relative_criterion_tolerance=1e-6,
+    convergence_scaled_gradient_tolerance=1e-6,
+    initial_radius=0.1,
+    population_size=100,
+)
+
+
+minimize(
+    # ...
+    algorithm=em.algorithms.scipy_lbfgsb,
+    algo_options=options,
+    # ...
+)
+```
+
+```{note}
+In my currently planned implementation, autocomplete will not work reliably for the
+copy constructors (`with_option`, `with_stopping` and `with_convergence`). The main
+reason is that most editors do not play well with `functools.wraps` or any other means
+of dynamic signature creation. For more details, see the discussions about the
+[Internal Algorithm Interface](algorithm-interface).
```

Comment:
Hyperlink doesn't seem to work in this environment.

Reply:
Syntax should be `[Internal Algorithm Interface](#algorithm-interface)`, I think.

Reply:
It works on [readthedcos](https://estimagic--495.org.readthedocs.build/en/495/development/eep-02-typing.html#id11) 

Which environment are you using?

Reply:
Just the [preview here on GitHub](https://github.com/OpenSourceEconomics/estimagic/blob/14c92cb028cae8083a390bd5ac0d2d972cbb1830/docs/source/development/eep-02-typing.md#proposal-4)